### PR TITLE
Save full sync status during enqueue

### PR DESCRIPTION
**Pull request by gravityrail**
For very large sites, it can take a long time to enqueue a full sync, and currently it needs to be done in the scope of a single request.

This PR changes a number of things about this:
- enqueues from highest to lowest ID, hopefully sending newer posts first in case everything crashes
- periodically saves the state of the current module while enqueuing (e.g. max_post_id for posts) so we can resume from that point if it crashes
- stops trying to enqueue the full sync if the total number of items in the full sync queue is greater than a certain (configurable) number. We already have a failsafe in place for this for queues in general, so I'm not 100% sure it's worth implementing this part, unless we expect the queue to grow SO dramatically beyond that.
- allows full sync enqueues to happen purely in request shutdown, not just in cron, which should help us relieve our dependence on (frequently broken) wp-cron implementations.

### CODE REVIEW

File: sync/class.jetpack-sync-module-full-sync.php [hunk 1]
```diff
@@ -196,55 +236,52 @@ public function get_status() {
...
+		$enqueue_status = $this->get_enqueue_status();
+		$module_config = $this->get_config();
+
 		foreach ( Jetpack_Sync_Modules::get_modules() as $module ) {
 			$name = $module->name();
 
-			if ( $total = $this->get_status_option( "{$name}_total" ) ) {
+			if ( ! isset( $module_config[ $name ] ) ) {
+				continue;
+			} else if ( $config = $module_config[ $name ] ) {
```
Comment (lezama) [inline sync/class.jetpack-sync-module-full-sync.php]: is the `else` needed?

File: json-endpoints/jetpack/json-api-jetpack-endpoints.php [hunk 2]
```diff
@@ -610,8 +610,6 @@
 		'full_queue_size' => '(int) Number of items in the full sync queue',
 		'full_queue_lag' => '(float) Time delay of the oldest item in the full sync queue',
 		'full_queue_next_sync' => '(float) Time in seconds before trying to sync the full sync queue again',
-		'is_scheduled' => '(bool) Is a full sync scheduled via cron?',
-		'cron_size'     => '(int) Size of the current cron array',
```
Comment (enejb) [inline json-endpoints/jetpack/json-api-jetpack-endpoints.php]: I think we should keep `cron_size` around.

File: json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php [hunk 3]
```diff
@@ -60,15 +60,9 @@ protected function result() {
...
 		$full_queue  = $sender->get_full_sync_queue();
 
-		$cron_timestamps = array_keys( _get_cron_array() );
-		$cron_age = microtime( true ) - $cron_timestamps[0];
-
 		return array_merge(
 			$sync_module->get_status(),
 			array(
-				'is_scheduled'          => Jetpack_Sync_Actions::is_scheduled_full_sync(),
-				'cron_size'             => count( $cron_timestamps ),
```
Comment (enejb) [inline json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php]: Lets keep cron_size and the cron age.

File: sync/class.jetpack-sync-actions.php [hunk 4]
```diff
@@ -165,85 +168,19 @@ static function schedule_initial_sync( $new_version = null, $old_version = null
 			$initial_sync_config['users'] = 'initial';
 		}
 
-		// we need this function call here because we have to run this function
-		// reeeeally early in init, before WP_CRON_LOCK_TIMEOUT is defined.
-		wp_functionality_constants();
-
-		if ( is_multisite() ) {
-			// stagger initial syncs for multisite blogs so they don't all pile on top of each other
-			$time_offset = ( rand() / getrandmax() ) * self::INITIAL_SYNC_MULTISITE_INTERVAL * get_blog_count();
```
Comment (enejb) [inline sync/class.jetpack-sync-actions.php]: Since we are deleting the `INITIAL_SYNC_MULTISITE_INTERVAL` we should also re move it every where in the file.

File: tests/php/sync/test_class.jetpack-sync-full.php [hunk 5]
```diff
@@ -598,14 +598,27 @@ function test_full_sync_sends_core_updates() {
...
 
 	function test_full_sync_start_sends_configuration() {
+		$post_ids = $this->factory->post->create_many( 3 );
+
 		// this is so that on WPCOM we can tell what has been synchronized in the past
 		add_action( 'jetpack_full_sync_start', array( $this, 'record_full_sync_start_config' ), 10, 1 );
 
+		$standard_config = array( 
+			'constants' => true,
+    		'functions' => true,
```
Comment (enejb) [inline tests/php/sync/test_class.jetpack-sync-full.php]: Indentation here is a bit off.

File: sync/class.jetpack-sync-module-full-sync.php [hunk 6]
```diff
@@ -188,49 +219,52 @@ public function get_status() {
...
-				$status[ 'config' ][ $name ] = $config;
-			}
 		}
 
 		return $status;
 	}
 
-	public function clear_status() {
-		$prefix = self::STATUS_OPTION_PREFIX;
+	public function clear_status() {
```
Comment (lezama) [inline sync/class.jetpack-sync-module-full-sync.php]: seems like we are introducing some unnecessary whitespace here

File: sync/class.jetpack-sync-actions.php (hunk 4 above)
Comment (gravityrail) [inline sync/class.jetpack-sync-actions.php]: fixed, thanks

File: json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php (hunk 3 above)
Comment (gravityrail) [inline json-endpoints/jetpack/class.jetpack-json-api-sync-endpoint.php]: I put it back in - thanks

File: json-endpoints/jetpack/json-api-jetpack-endpoints.php (hunk 2 above)
Comment (gravityrail) [inline json-endpoints/jetpack/json-api-jetpack-endpoints.php]: fixed

**Comment by lezama** [comment]
I am merging this, code looks 👍, works on my testings, and better to have it tested early by others than later :)

great job @gravityrail! Thanks 🙇
